### ISSUE TITLE
E722 considered harmful

### ISSUE THREAD

**Issue by couling** [PyCQA/pycodestyle#1009]
I realise that #703 already got locked as "too heated".  But I must ask that this issue be revised carefully and cooly to come to some sort of a solution.  In its current form E722 is more dangerous than it is helpful.

I would like to put it on record that the decision to leave that as-is has tripped me, my colleagues, and many others into writing some serious bugs into our code.  Initially I felt pretty embarrassed for not knowing about this for so long, but the more I've talked to both colleagues and friends about this the more I've realised that developers are often not aware of the bug it introduces:

```pycharm
# The user hit's control-C during do_something() and it left x in a really bad state
# This was supposed to be caught by cleanup()
# So now the user has seriously corrupted data
try:
    do_something(x)
except Exception:  # Really this should have been just except:
    cleanup(x)
    raise
```

PEP 8 does not ban the use of bare `except:`.  That's plainly explicit in PEP 8.  It calls out two specific and common use cases where it is permissable.  I say "common" because logging is not an edge case requirement.  And I also say "common" because resource cleanup is not an edge case requirement and cannot always be done in a `finally:` block.

-----

Most novices and even many intermediate developers I've spoken to will be to swap `except:` for `except Exception:` where their code was already re-raising the exception with a bare `raise`.  They do this because they think PyCQA's E722 (via Pycharm or similar) has told them to.  When they make this change the warning goes away.  But what have they done? 

- has they fixed a problem? **No** not really.  If their code was capturing the exception and suppressing it (not re-raising) then `except Exception:` is almost always too broad.  Their code is just as likely to misbehave on a badly formatted date or a function call with too few arguments.  It's only fixed one small specific problem of the user hitting ctrl-c.  If their code was already re-raising then there was no bug to start with and this has fixed nothing at all.
  
-  Has it introduced a new bug? **Yes** If they really did intend to catch everything then nudging them away from `except:` onto `except Exception:` simply will not do what they intend.  But worse it will look like it does what they intend long enough to get into production many times over.

The potential bugs PEP 8 / E722 are concerned with are all to do with _suppressing_ exceptions incorrectly.  If PyCQA is incapable of analysing suppression (incapable of seeing the future `raise) this then it's incapable of applying this rule well enough to avoid very many dangerous false positives.

----

In conclusion, I can find nearly no instances where E722 has practically nudged myself or colleagues into writing better code with fewer bugs.  I can find many of instances where it has nudged myself or colleagues into writing faulty code.

I now believe this "bare except is always bad" interpretation of PEP 8 is much more harmful than it is ever helpful.

**Comment by asottile** [comment]
dupe #703
